### ISSUE TITLE
CUDA/HIP graph capture of the fissioned step (generalize Metal's fused segment encoding)

### ISSUE THREAD

**Issue by lukstafi** [ahrefs/ocannl#488]
Metal's fused segment encoding (`sequence_segments`: one command-buffer callback encoding a whole serial chain of fissioned steps) was the single biggest training-throughput win on small models — it attacked per-dispatch overhead, not kernel quality. CUDA and HIP have no counterpart: every fissioned segment is a separate kernel launch each step, and for `lenet`/`mlp_small`-class models at small batch, launch overhead is a first-order cost. tinygrad and torch both lean on CUDA Graphs for exactly this.

## The idea

Capture the fissioned step's launch sequence as a **CUDA Graph** (`cuGraphCreate`/`cuGraphAddKernelNode` explicitly, or stream-capture around the existing launch loop) and replay it with `cuGraphLaunch` — one API call per training step instead of one per segment. HIP mirrors the API (`hipGraph*`) nearly 1:1, so the design should live in shared backend logic with thin driver bindings.

The structure is already there: the fission segments give a static launch sequence with known buffers, and the "same routine back-to-back" execution pattern that motivated Metal's FIFO-wait fix is precisely the graph-replay pattern. The Metal `sequence_segments` hook defines the seam — this issue generalizes that seam to a per-backend "encode a segment chain" capability rather than a Metal special case.

## Known constraints

- **Stable arguments**: graph nodes bake kernel params; OCANNL's routines already have stable buffer bindings per context, and merge-buffer/pool pointers must be pinned for the graph's lifetime — re-instantiate (`cuGraphExecUpdate`) on context changes rather than rebuilding.
- **Host callbacks**: `debug_log_from_routines` and host-side logging don't belong in a graph — keep the unfused launch path for logged runs (the same rule the Metal encoding and the tensorized renderings already follow).
- **Events/streams**: cross-stream waits inside a step become graph dependencies; the existing event machinery maps onto graph edges.
- **cudajit/hipjit surface**: needs new bindings (graph create/instantiate/launch/exec-update) — coordinate with the gh-482 pin bump so it's one cudajit release, not two.

## Tasks

1. Generalize the segment-chain encoding seam from Metal-specific to `Backend_intf` (capability-gated).
2. cudajit + hipjit graph bindings.
3. CUDA implementation via stream capture of the existing launch loop (lowest-risk first cut); measure on `benchmarks/` small-model workloads.
4. Fall back transparently (per-segment launches) when capture is unsupported or logging is on.

Disjoint from: kernel-quality issues (gh-479/480/481); this is pure runtime. Related: gh-482 (shared cudajit release), gh-160 (tracing overhead — same "host-side cost per step" family).

**Comment by lukstafi** [comment]
## Measured: what graph replay is actually worth on this hardware

RTX 5070 Ti Laptop (sm_120), driver 610.62 / CUDA 13.3, WSL2. Done as an acceptance-target
measurement for this issue, from the gh-ocannl-476 CUDA leg.

### The probe

A standalone CUDA program (no OCANNL) timing a chain of N no-op kernels — the shape of a fissioned
step — three ways: N individual launches + one sync; the same chain stream-captured into a graph and
replayed; and async enqueue with the sync outside the timed region. Median of 200 reps after 50
warmup. Source: [`launchprobe.cu` gist-able from the sweep scratch, reproduced below].

| N | N launches + sync | graph + sync | saved/step | per launch | async enqueue/launch |
|---|---|---|---|---|---|
| 8 | 51.9 µs | 19.4 µs | **32.5 µs** | 6.49 µs | 2.78 µs |
| 16 | 119.3 µs | 25.6 µs | **93.7 µs** | 7.46 µs | 2.65 µs |
| 24 | 159.1 µs | 28.7 µs | **130.5 µs** | 6.63 µs | 2.81 µs |
| 36 | 233.5 µs | 37.9 µs | **195.6 µs** | 6.49 µs | 2.68 µs |
| 64 | 411.7 µs | 49.1 µs | **362.6 µs** | 6.43 µs | 2.65 µs |
| 117 | 735.4 µs | 82.9 µs | **652.5 µs** | 6.29 µs | 2.62 µs |

So on this box: **~6.4 µs per individual launch, versus a graph replay costing ~15 µs fixed plus
~0.58 µs per node.** Graphs are worth it from about N=3 upward, and the marginal saving is ~5.8 µs
per segment. Host-side enqueue alone is 2.65 µs, so roughly 3.7 µs of the 6.4 is device-side
inter-kernel gap that only graph replay (not CPU run-ahead) can remove.

### Applied to the benchmark workloads

Launches per step: from the `bench_conv_diag` / `bench_gpt_diag` census for the conv and gpt
workloads, and from the autotune calibration line (`... N kernels`) for `mlp_wide`. `mlp_small` has
neither, so its 13 is counted from generated `__global__` definitions — the one soft number here
(that method gives 16 for `mlp_wide` against a true 15, so treat it as +1 or so).

| workload | variant | launches/step | measured step | est. saved | share of step |
|---|---|---|---|---|---|
| **mlp_small** | default | ~13 | 0.112 ms | ~71 µs | **~63%** |
| mlp_wide | default | 15 | 4.167 ms | ~86 µs | 2.1% |
| lenet | tuned | 35 | 11.560 ms | ~190 µs | 1.6% |
| cifar_stride | tuned | 36 | 31.113 ms | ~196 µs | 0.6% |
| gpt2_mini | default | 117 | 188.905 ms | ~652 µs | 0.3% |
| cifar_conv | materialized | 36 | 88.991 ms | ~196 µs | 0.2% |
| lenet | default | 24 | 242.612 ms | ~130 µs | 0.1% |

### Reading

**The issue's premise holds, but only for the smallest workload.** `mlp_small` is genuinely
dispatch-bound — ~63% of its 112 µs step is launch overhead that graph replay removes. Corroborating
that independently: its `queued_step_ms` is 99 µs, i.e. even with the per-step sync amortized away
the step still costs ~7.6 µs per kernel, which is the inter-kernel gap rather than GPU work.

**`lenet` is not in that regime**, despite being named alongside `mlp_small` in the issue: 1.6% on
the tuned cell, 0.1% on the default. Its segments do enough real work to dwarf 6 µs each.

**And it is not the lever for `gpt2_mini`** (0.3%), which is what I measured separately in
gh-ocannl-531 — there the step is 85% matmul in nine dominant segments, blocked by gh-ocannl-521.

One caveat on the value of the `mlp_small` win: at 0.112 ms it is *already* the fastest cell in that
workload across all three frameworks (torch cuda 0.620 ms, tinygrad CUDA 0.121 ms). So this is
extending a lead rather than closing a gap — real, but low stakes. The gaps that matter in the
matrix (convs at 20–74x, `gpt2_mini` at 72x) are untouched by graph capture.

### Acceptance target for when this lands

From the table, a correct implementation should show, on CUDA:

- `mlp_small/cuda` default: **0.112 ms → ~0.04–0.05 ms**, the headline check.
- `mlp_wide/cuda` default: ~4.167 → ~4.08 ms (2%, near run-to-run noise — needs several rounds to
  call).
- `lenet/cuda` tuned: ~11.56 → ~11.37 ms.
- Everything else: no measurable change; treat a regression anywhere as a bug in the fallback path.

If the achieved saving materially *exceeds* these, that is informative too — it would mean OCANNL's
per-segment host bookkeeping (event handling, binding updates) costs more than a bare
`cuLaunchKernel`, which the probe does not capture.

Caveats: WSL2 adds some driver-call overhead versus native Linux, so the per-launch figure may be
pessimistic for other hosts; and the probe's kernels are empty, so on real segments some dispatch
latency can overlap with execution, making these estimates upper bounds on a per-workload basis.

<details><summary>probe source</summary>

```cuda
__global__ void nop() {}
// per N: (a) N launches then cudaStreamSynchronize; (b) stream-capture the same chain,
// cudaGraphInstantiate, then cudaGraphLaunch + sync; (c) N launches, sync outside the timed region.
// median of 200 reps after 50 warmup, kernels launched <<<8,128>>> on one non-default stream.
```
</details>

**Comment by lukstafi** [comment]
Implemented in lukstafi/ocannl-staging#276 (cudajit 0.8.0 / hipjit 0.2.0 `Graph` modules landed on their `main` branches). Against the acceptance targets above, measured on the RTX 5070 Ti with `--ocannl_gpu_graph_capture=false` as the same-binary control (p50 step / queued step, ms):

- **mlp_small/cuda default: 0.105 → 0.055** (queued 0.099 → 0.040) — inside the predicted 0.04–0.05 window.
- mlp_small/hip: 0.271 → 0.133 (queued 0.222 → 0.086).
- mlp_wide/hip: 1.99 → 1.79 — the per-launch overhead on the iGPU is high enough that even the GEMM-bound workload gains ~10%.
- mlp_wide/cuda: queued 4.077 → 4.035 (~1%, as predicted); sync-after-every-step p50 shows a +3% graph re-scheduling bubble on the idle device, which pipelined execution hides — `cuGraphUpload` prefetch noted as a follow-up.
- lenet/cuda default: unchanged (114.7 → 114.6), consistent with the 0.1% estimate.

Loss trajectories are bitwise identical with capture on vs off. One implementation surprise worth recording: capture must use the RELAXED mode, because GC finalizers (dead-module unloads) fire mid-capture on the capturing thread and stricter modes turn them into capture-invalidating errors that escape `Gc.finalise` at arbitrary program points.

**Comment by lukstafi** [comment]
Closing on lukstafi/ocannl-staging#276, merged 2026-08-04 (with cudajit 0.8.0 / hipjit 0.2.0 shipping the `Graph` modules). The acceptance targets are met per the measurement comment above: mlp_small/cuda 0.105 → 0.055 ms p50 (inside the predicted window), mlp_small/hip 2x, mlp_wide/hip ~10%, and the launch-bound-only prediction confirmed by lenet's unchanged 114.7 → 114.6 and mlp_wide/cuda's ~1%. Loss trajectories bitwise identical with capture on vs off.

Recorded follow-up, not worth an open issue unless it starts mattering: `cuGraphUpload` prefetch to hide the +3% graph re-scheduling bubble visible only under sync-after-every-step on an idle device (pipelined execution already hides it). The RELAXED-capture-mode requirement (GC finalizers firing mid-capture) is documented in the implementation.
